src/core/src/signature.rs

```diff
@@ -163,27 +164,149 @@ impl SigsTrait for Sketch {
     }
 }
 
-// Iterator for converting sequence to hashes
+#[derive(Debug, Clone)]
+pub enum ReadingFrame {
+    DNA {
+        fw: Vec<u8>,
+        rc: Vec<u8>,
+        len: usize, // len gives max_index for kmer iterator
+    },
+    Protein {
+        fw: Vec<u8>,
+        len: usize,
+    },
+}
+
+impl std::fmt::Display for ReadingFrame {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ReadingFrame::DNA { fw, rc, len } => {
+                let fw_str = String::from_utf8_lossy(fw).to_string();
+                let rc_str = String::from_utf8_lossy(rc).to_string();
+                write!(
+                    f,
+                    "Type: DNA ({}bp), Forward: {}, Reverse Complement: {}",
+                    len, fw_str, rc_str
+                )
+            }
+            ReadingFrame::Protein { fw, len } => {
+                let fw_str = String::from_utf8_lossy(fw).to_string();
+                write!(f, "Type: Protein ({}aa), Forward: {}", len, fw_str)
+            }
+        }
+    }
+}
+
+impl ReadingFrame {
+    pub fn new_dna(sequence: &[u8]) -> Self {
+        let fw = sequence.to_ascii_uppercase();
+        let rc = revcomp(&fw);
+        let len = sequence.len();
+        ReadingFrame::DNA { fw, rc, len }
+    }
+
+    pub fn new_protein(sequence: &[u8], dayhoff: bool, hp: bool) -> Self {
+        let seq = sequence.to_ascii_uppercase();
+        let fw: Vec<u8> = if dayhoff {
+            seq.iter().map(|&aa| aa_to_dayhoff(aa)).collect()
+        } else if hp {
+            seq.iter().map(|&aa| aa_to_hp(aa)).collect()
+        } else {
+            seq
+        };
+
+        let len = fw.len();
+        ReadingFrame::Protein { fw, len }
+    }
+
+    pub fn new_skipmer(sequence: &[u8], start: usize, m: usize, n: usize) -> Self {
+        let seq = sequence.to_ascii_uppercase();
+        if start >= n {
+            panic!("Skipmer frame number must be < n ({})", n);
```

Comment:
Should we `panic!` here and below on `new_translated`, or change the fn signature to return a `Result`?

Reply:
Would a result be cleaner so we can propagate the error? 

cc @ctb

Reply:
implemented in #3445, would appreciate a review there